src/updater.rs

```diff
@@ -0,0 +1,245 @@
+    let conns = crate::Connection::alive_conns();
+    conns.is_empty() && has_no_controlling_conns()
+}
+
+#[cfg(any(not(target_os = "windows"), feature = "flutter"))]
+fn has_no_controlling_conns() -> bool {
+    CONTROLLING_SESSION_COUNT.load(Ordering::SeqCst) == 0
+}
+
+#[cfg(not(any(not(target_os = "windows"), feature = "flutter")))]
+fn has_no_controlling_conns() -> bool {
+    let app_exe = format!("{}.exe", crate::get_app_name().to_lowercase());
+    for arg in [
+        "--connect",
+        "--play",
+        "--file-transfer",
+        "--view-camera",
+        "--port-forward",
+        "--rdp",
+    ] {
+        if !crate::platform::get_pids_of_process_with_first_arg(&app_exe, arg).is_empty() {
+            return false;
+        }
+    }
+    true
+}
+
+fn start_auto_update_check() -> Sender<UpdateMsg> {
+    let (tx, rx) = channel();
+    std::thread::spawn(move || start_auto_update_check_(rx));
+    return tx;
+}
+
+fn start_auto_update_check_(rx_msg: Receiver<UpdateMsg>) {
+    std::thread::sleep(Duration::from_secs(30));
+    if let Err(e) = check_update(false) {
+        log::error!("Error checking for updates: {}", e);
+    }
+
+    const MIN_INTERVAL: Duration = Duration::from_secs(60 * 10);
+    const RETRY_INTERVAL: Duration = Duration::from_secs(60 * 30);
+    let mut last_check_time = Instant::now();
+    let mut check_interval = DUR_ONE_DAY;
+    loop {
+        let recv_res = rx_msg.recv_timeout(check_interval);
+        match &recv_res {
+            Ok(UpdateMsg::CheckUpdate) | Err(_) => {
+                if last_check_time.elapsed() < MIN_INTERVAL {
+                    // log::debug!("Update check skipped due to minimum interval.");
+                    continue;
+                }
+                // Don't check update if there are alive connections.
+                if !has_no_active_conns() {
+                    check_interval = RETRY_INTERVAL;
+                    continue;
+                }
+                if let Err(e) = check_update(matches!(recv_res, Ok(UpdateMsg::CheckUpdate))) {
+                    log::error!("Error checking for updates: {}", e);
+                    check_interval = RETRY_INTERVAL;
+                } else {
+                    last_check_time = Instant::now();
+                    check_interval = DUR_ONE_DAY;
+                }
+            }
+            Ok(UpdateMsg::Exit) => break,
+        }
+    }
+}
+
+fn check_update(manually: bool) -> ResultType<()> {
+    #[cfg(target_os = "windows")]
+    let is_msi = crate::platform::is_msi_installed()?;
+    if manually || config::Config::get_bool_option(config::keys::OPTION_ALLOW_AUTO_UPDATE) {
+        if do_check_software_update().is_ok() {
+            let update_url = crate::common::SOFTWARE_UPDATE_URL.lock().unwrap().clone();
+            if update_url.is_empty() {
+                log::debug!("No update available.");
+            } else {
+                let download_url = update_url.replace("tag", "download");
```

Comment:
Too many unnessary
{
{
{

here all you can use  `if .. return instead`

Reply:
Applied.